Documentation/performance/JitOptimizerPlanningGuide.md

```diff
@@ -0,0 +1,127 @@
+   of the measurements, which will usually be at least 3% or so even for the
+   most well-behaved macro-benchmarks.
+ - Source-level changes can be made much more rapidly than compiler optimization
+   changes.  This means that for anything we're trying to track where the whole
+   team is effecting changes in source, runtime, etc., any particular code
+   sequence we may target with optimization improvements may well be targeted
+   with source changes in the interim, nullifying the measured benefit of the
+   optimization change when it is eventually merged.  Source/library/runtime
+   changes are in play for TechEmpower and Benchmarks Game both.
+
+Compiler micro-benchmarks (like those in our [test tree](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality))
+don't share these issues, and adding them as optimizations are implemented is
+critical for validation and regression prevention; however, micro-benchmarks
+often aren't as representative of real-world code, and therefore not as
+reflective of developers' performance needs, so aren't well suited for scouting
+out and prioritizing opportunities.
+
+
+Benefits of JIT Optimization
+----------------------------
+
+While source changes can more rapidly and dramatically effect changes to
+targeted hot code sequences in macro-benchmarks, compiler changes have the
+advantage that they apply broadly to all compiled code.  One of the best reasons
+to invest in compiler optimization improvements is to capitalize on this.  A few
+specific benefits:
+
+ - Optimizer changes can effect "peanut-butter" improvements; by making an
+   improvement which is small in any particular instance to a code sequence that
+   is repeated thousands of times across a codebase, they can produce substantial
+   cumulative wins.  These should accrue toward the standard metrics (benchmark
+   scores and code size), but identifying the most profitable "peanut-butter"
+   opportunities is difficult.  Improving our methodology for identifying such
+   opportunities would be helpful; some ideas are below.
+ - Optimizer changes can unblock coding patterns that performance-sensitive
+   developers want to employ but consider prohibitively expensive.  They may
+   have inelegant works-around in their code, such as gotos for loop-exiting
+   returns to work around poor block layout, manually scalarized structs to work
+   around poor struct promotion, manually unrolled loops to work around lack of
+   loop unrolling, limited use of lambdas to work around inefficient access to
+   heap-allocated closures, etc.  The more the optimizer can improve such
+   situations, the better, as it both increases developer productivity and
+   increases the usefulness of abstractions provided by the language and
+   libraries.  Finding a measurable metric to track this type of improvement
+   poses a challenge, but would be a big help toward prioritizing and validating
```

Comment:
We use all of those, including also:
- Forcing inlining like crazy to avoid methods calls 
- Using goto's to build "sub-routines" avoiding the cost of the call and the size of the multiple instances of inlining on big methods. 
Examples of the result: https://ayende.com/blog/177569/why-we-arent-publishing-benchmarks-for-ravendb-4-0-yet